### DISCUSSION TITLE
Can I provide a predefined node mapping between two graphs to guide or constrain the isomorphism process?

### DISCUSSION THREAD

**Discussion by Mohsin-Bashir** [networkx/networkx#7734] [Q&A] [answered]
I am using GM.isomorphisms_iter() to generate all possible isomorphisms between two graphs with 60 nodes each. However, this process is extremely slow. Is there a way to provide a predefined node mapping to speed up the computation or prioritize certain isomorphisms?

**Comment by dschult** [comment]
Can you explain more about what you mean by a “predefined node mapping”?  
If you map the nodes between two graphs, then you have an isomorphism.  
Are you wanting to prescribe a partial mapping to start from? 
Or is your predefined mapping different than what we mean by isomorphism mapping?
Can you give an example, or more info about what you are actually trying to do in the end?

**Comment by Mohsin-Bashir** [comment]
Suppose I have two graphs G1 and G2, each with nodes labeled {1,2,3,…,60} and I know that:
•	Node 1 in G1 should map to Node 45 in G2.
•	Node 2 in G1 should map to Node 30 in G2.
If this is my predefined mapping I want to restrict the search space to only consider isomorphisms that include these mappings. Otherwise, the process takes too long to generate all the isomorphisms. Ultimately, my goal is to determine the possible combinations of node mappings from G1 to G2.

**Reply by dschult** [reply]
Ahh.. I see.  It almost sounds like you are reducing the problem to a smaller one. 

And I guess we should assume that these choices have already been passed through the checks as valid choices. You might be able to use methods from GM to make sure they are feasible choices.

Right now, the best way to do this is to find the isomorphism between the smaller graphs without those nodes. Making the problem smaller will help with speed as well. 
```python
G1small = G1.copy()
G1small.remove_nodes_from([1, 2])
G2small = G2.copy()
G2small.remove_nodes_from([45, 30])

< find mapping m between G1small and G2small >
m_full = m.update({1: 45, 2:30})
```

It is probably possible to figure out how to set up the inner storage of GM to represent "starting" from a specified mapping. But it might be tricky to get right for all the corner cases. And we don't have that feature currently in NetworkX.

**Reply by boothby** [reply, marked as answer]
I don't think that's quite right, Dan.  That approach can produce psudo-isomorphisms that interchange the roles of 1 and 2, for example.  This gets worse as the number of "pinned" nodes grows.

However, you can add colors to your graph, and tell the GraphMatcher to respect the coloring.

```python
G1small.nodes[1]['color'] = 0
G1small.nodes[2]['color'] = 1
G2small.nodes[45]['color'] = 0
G2small.nodes[30]['color'] = 1
# note: all other nodes are left with color None

def color_match(node_attr_1, node_attr_2):
    return node_attr_1.get('color') == node_attr_2.get('color')

GM = GraphMatcher(G1small, G2small,  node_match=color_match)
```

edit: one needn't define a function, it seems that `node_match = categorical_node_match('color', None)` would suffice:
https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isomorphism.categorical_node_match.html

**Reply by dschult** [reply]
Nice!  Thanks @boothby This is a much better approach.

**Reply by rossbar** [reply]
Generally speaking this use-case - i.e. the "constrained" isomorphism problem where part of the mapping is known a priori - seems like a potentially interesting feature. I haven't looked at the literature, but I wonder whether/how well this could be supported in the other isomorphism functions (vf2++, etc.)...

**Reply by Mohsin-Bashir** [reply]
Thank you so much @boothby. I shall try this approach.
